### PR TITLE
k8s: Split SharedResources into binary specific cells

### LINKED ISSUE #24070: Modularization: split SharedResources into OperatorResources and DaemonResources
SharedResourcesCell provides a set of shared handles to Kubernetes resources.
At the moment, the same cell is used in the daemon, the operator and the clustermesh-apiserver, despite each binary does not use all the resources.

To avoid having a single cell with too many unrelated resources, we should split the SharedResourcesCell into separate ones specific for each binary. This will help to understand which resources each binary is actually using.

**Success Criteria**

- The SharedResourcesCell is split in more binary-specific cell

Related: #23425

### PR DESCRIPTION
**Pull request by pippolo84**
Split the SharedResources cell into multiple cells, one for each binary (agent, operator and clustermesh-apiserver).
This improves the dependencies management, since each binary will depend only on the k8s resources that it actually uses.

Also, the functional options pattern is introduced to give more flexibility when customizing a resource events stream.

_Notes for the reviewers_: review commit by commit might be simpler. The last commit is the one responsible for the high number of files modified, but it is just a rename (`sharedResources` -> `resources`) after the removal of the single SharedResources cell.

Fixes https://github.com/cilium/cilium/issues/24070

### CODE REVIEW

File: pkg/k8s/resource_ctors.go
Comment (joamaki) [inline pkg/k8s/resource_ctors.go]: I wonder if it'd make sense to namespace these by package, e.g. `pkg/k8s/resources` and then we'd have `resources.CiliumNetworkPolicy`, or by the suffix (`NamespaceResource`). I think either is probably fine and I do like having `k8s` in the package name. If you go the suffix route, then let's add it to these `Cilium*` functions here.

Comment (pippolo84) [reply pkg/k8s/resource_ctors.go]: Added the `Resource` suffix where missing.

**Review by lmb — APPROVED** [review]

**Review by nebril — APPROVED** [review]

**Review by gandro — APPROVED** [review]
My codeowners look good to me.

It might be a naive question, but why do we even need the per-binary packages? Can't there be something like `cell.LazyProvide` that only executes the constructor if it's actually requested by a dependency?

**Comment by pippolo84** [comment]
> It might be a naive question, but why do we even need the per-binary packages? Can't there be something like `cell.LazyProvide` that only executes the constructor if it's actually requested by a dependency?

I think it'd be a reasonable alternative for the current usage of the resources, even if keeping a single cell would not allow to customize each resource events stream when consumed by multiple binaries.
As an example, we might want to use different selectors (or different [TransformFunc](https://pkg.go.dev/k8s.io/client-go@v0.26.0/tools/cache#TransformFunc) when we'll add support for that in resource.Resource[T]).

**Review by christarazi — APPROVED** [review]

File: clustermesh-apiserver/k8s/resources.go
Comment (YutaroHayakawa) [inline clustermesh-apiserver/k8s/resources.go:22]: I think we need CiliumNode, Endpoints, and CiliumEndpoint resources, but they are currently using the legacy `clientset`. Approving with the understanding that they will be followed up at some point.

**Review by YutaroHayakawa — APPROVED** [review]
Looks good to me from ClusterMesh's perspective.

File: clustermesh-apiserver/k8s/resources.go
Comment (pippolo84) [reply clustermesh-apiserver/k8s/resources.go:22]: Yep, there is already work in progress to port other resources to the new `resource.Resource[T]` framework.

**Comment by pippolo84** [comment]
Force-pushed to solve a conflict in `clustermesh-apiserver/main.go`

**Review by ldelossa — APPROVED** [review]
LGTM

**Pull request merged**
